### PR TITLE
feat: Node.js server SDK first version

### PR DESCRIPTION
**Pull request by cosmos-explorer**

### CODE REVIEW

File: src/FbClientCore.ts [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    defaultValue: TValue,
+    typeConverter: (value: string) => IConvertResult<TValue>
+  ): IEvalDetail<TValue> {
+    if (!this.initialized()) {
+      this.logger?.warn(
+        'Variation called before FeatBit client initialization completed (did you wait for the' +
+        "'ready' event?) - using default value",
+      );
+
+      return {kind: ReasonKinds.ClientNotReady, reason: 'client not ready', value: defaultValue};
```
Comment (deleteLater) [inline src/FbClientCore.ts:299]: Maybe `ReasonKinds.InvalidContext`

File: src/evaluation/collection.ts
Comment (deleteLater) [inline src/evaluation/collection.ts:1]: unused file

File: src/evaluation/IEvalDetail.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { ReasonKinds } from "./ReasonKinds";
+
+export interface IEvalDetail<TValue> {
+  /**
+   * The result of the flag evaluation. This will be either one of the flag's variations or
+   * the default value that was passed to `FbClient.variationDetail`.
+   */
```
Comment (deleteLater) [inline src/evaluation/IEvalDetail.ts]:
```suggestion
  /** 
  * An enum indicating the category of the reason.
  */
```

File: src/evaluation/Evaluator.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+/**
+ * @internal
+ */
+export default class Evaluator {
+  constructor(private platform: IPlatform, private store: IStore) {
+  }
+
+  /**
+   * Evaluate the given flag against the given context.
+   * @param flag The flag to evaluate.
```
Comment (deleteLater) [inline src/evaluation/Evaluator.ts]:
```suggestion
   * @param feature flag key
```

File: src/evaluation/Evaluator.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    return this.evaluateRules(flag, context);
+  }
+
+  /**
+   * Evaluate the targets of the specified flag against the given context.
+   * @param flag The flag to evaluate targets for.
+   * @param context The context to evaluate those targets against.
+   * @returns An evaluation result if there is a target match/error or undefined if there is not.
```
Comment (deleteLater) [inline src/evaluation/Evaluator.ts]: remove or change `@returns` comment

File: src/evaluation/Evaluator.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+   * Attempt to get an evaluation result for the specific variation/flag combination.
+   * @param flag The flag to get a variation from.
+   * @param index The index of the flag.
+   * @param reason The initial evaluation reason. If there is a valid variation, then this reason
+   * will be returned in the EvalResult.
+   * @returns An evaluation result containing the successful evaluation, or an error if there is
+   * a problem accessing the variation.
+   *
+   * @internal
+   */
```
Comment (deleteLater) [inline src/evaluation/Evaluator.ts]: 1. remove comment or change
2. do we need this `getVariation` method?

File: src/evaluation/Evaluator.ts
Comment (deleteLater) [inline src/evaluation/Evaluator.ts:1]: A lot of wrong comments found in this file, WebStorm highlight them


image

File: src/evaluation/operator.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+  public static readonly NotOneOf: Operator =
+    new Operator(OperatorTypes.NotOneOf, (param1: string, param2: string) => {
+      try {
+        const values: string[] = JSON.parse(param2);
+        return values.includes(param1);
+      } catch (e) {
+        return false;
+      }
+    });
```
Comment (deleteLater) [inline src/evaluation/operator.ts]: BUG

File: src/evaluation/operator.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+  /*********************** is true/ is false *******************************/
+  public static readonly IsTrue: Operator =
+    new Operator(OperatorTypes.IsTrue, (param1: string, _: string) => Regex.fromString('/^true$/i').test(param1));
+
+  public static readonly IsFalse: Operator =
+    new Operator(OperatorTypes.IsTrue, (param1: string, _: string) => Regex.fromString('/^false$/i').test(param1));
```
Comment (deleteLater) [inline src/evaluation/operator.ts]: Case-insensitive string comparison: https://masteringjs.io/tutorials/fundamentals/compare-strings-ignore-case

File: src/events/DefaultEventSender.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    } catch (err) {
+      res.status = DeliveryStatus.Failed;
+      res.error = err;
+      return res;
+    }
+
+    // wait 1 second before retrying
+    await sleep();
+
+    return this.send(payload);
```
Comment (deleteLater) [inline src/events/DefaultEventSender.ts]: When will retry ended

File: src/FbClientCore.ts [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+      return {kind: evalResult.kind, reason: evalResult.reason, value: defaultValue};
+    }
+
+    if (evalResult.kind === ReasonKinds.FlagNotFound) {
+      // flag not found, return default value
+      const error = new ClientError(evalResult.reason!);
+      this.onError(error);
+
+      return {kind: evalResult.kind, reason: evalResult.reason, value: defaultValue};
+    }
```
Comment (deleteLater) [inline src/FbClientCore.ts]: duplicated code

File: src/options/IOptions.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+   * The base URI of the event service
+   */
+  eventsUri?: string;
+
+  /**
+   * Whether streaming mode should be used to receive flag updates.
+   *
+   * This is true by default. If you set it to false, the client will use polling.
+   */
+  stream?: boolean;
```
Comment (deleteLater) [inline src/options/IOptions.ts]: How about add a field called `mode`

Comment (cosmos-explorer) [reply src/options/IOptions.ts]: done

File: src/FbClientCore.ts (hunk 9 above)
Comment (cosmos-explorer) [reply src/FbClientCore.ts]: done

File: src/evaluation/operator.ts (hunk 7 above)
Comment (cosmos-explorer) [reply src/evaluation/operator.ts]: done

File: src/evaluation/operator.ts (hunk 6 above)
Comment (cosmos-explorer) [reply src/evaluation/operator.ts]: done

File: src/evaluation/Evaluator.ts
Comment (cosmos-explorer) [reply src/evaluation/Evaluator.ts:1]: done

File: src/evaluation/Evaluator.ts (hunk 5 above)
Comment (cosmos-explorer) [reply src/evaluation/Evaluator.ts]: done

File: src/evaluation/Evaluator.ts (hunk 4 above)
Comment (cosmos-explorer) [reply src/evaluation/Evaluator.ts]: done

File: src/evaluation/Evaluator.ts (hunk 3 above)
Comment (cosmos-explorer) [reply src/evaluation/Evaluator.ts]: done

File: src/evaluation/IEvalDetail.ts (hunk 2 above)
Comment (cosmos-explorer) [reply src/evaluation/IEvalDetail.ts]: done

File: src/evaluation/collection.ts
Comment (cosmos-explorer) [reply src/evaluation/collection.ts:1]: done

File: src/events/DefaultEventSender.ts (hunk 8 above)
Comment (cosmos-explorer) [reply src/events/DefaultEventSender.ts]: retry once

File: src/Context.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  get key(): string {
+    return this._user!.key;
+  }
+
+  value(property: string): any {
+    const propertyName = property as keyof IUser;
+    return this._user?.[propertyName] ?? this._user?.customizedProperties?.find(x => x.name === propertyName)?.value;
+  }
```
Comment (deleteLater) [inline src/Context.ts]: BUG

```csharp
public string ValueOf(string property)
{
    if (string.IsNullOrWhiteSpace(property))
    {
        return string.Empty;
    }

    if (property == "keyId")
    {
        return Key;
    }

    if (property == "name")
    {
        return Name;
    }

    return Custom.TryGetValue(property, out var value) ? value : string.Empty;
}
```

**Review by deleteLater — APPROVED** [review]
LGTM

**Pull request merged**